src/common/transformations/src/transformations/mlir/mlir_op.cpp

```diff
@@ -267,6 +290,138 @@ namespace mlir {
+    gc::gpu::StaticExecutor exec(module);
+
+    auto it = evaluationContext.find(ov::internal::mlir_meta::is_kernel_arg_usm.name());
+    if (it == evaluationContext.end()) {
+        OPENVINO_THROW("No is_kernel_arg_usm provided for OpenCL execution");
+    }
+    std::vector<bool> arg_types = it->second.as<std::vector<bool>>();
+
+    for (size_t i = 0; i < inputs.size(); ++i) {
+        exec.arg(inputs[i].data(), arg_types[i]);
+    }
+    for (size_t i = 0, j = inputs.size(); i < outputs.size(); ++i, ++j) {
+        exec.arg(outputs[i].data(), arg_types[j]);
+    }
+    exec(ctx);
+    maybe_set_result_event(evaluationContext, ctx);
+    return true;
+}
+
+bool MLIREvaluateGcGPU::invoke_packed(std::vector<void*>& args, const ov::EvaluationContext& evaluationContext) {
+    gc::gpu::OclContext ctx = build_ocl_context(evaluationContext);
+    gc::gpu::DynamicExecutor exec(module);
+
+    auto it = evaluationContext.find(ov::internal::mlir_meta::is_kernel_arg_usm.name());
+    if (it == evaluationContext.end()) {
+        OPENVINO_THROW("No is_kernel_arg_usm provided for OpenCL execution");
+    }
+    std::vector<bool> argTypes = it->second.as<std::vector<bool>>();
+    for (size_t i = 0; i < args.size(); i+=4) {
+        exec.arg(
+            /*alignedPtr=*/args[i],
+            /*rank=*/reinterpret_cast<size_t>(args[i + 1]),
+            /*shape=*/reinterpret_cast<int64_t*>(args[i + 2]),
+            /*strides=*/reinterpret_cast<int64_t*>(args[i + 3]),
+            /*isUsm=*/argTypes[i]
+        );
+    }
+    exec(ctx);
+    maybe_set_result_event(evaluationContext, ctx);
+    return true;
+}
+
+void MLIREvaluateGcGPU::maybe_set_result_event(const ov::EvaluationContext& evaluationContext, gc::gpu::OclContext& ctx) {
+    // case with in-order queue where we don't need to return an event
+    if (ctx.lastEvent == nullptr)
+        return;
+    auto it = evaluationContext.find(ov::internal::mlir_meta::result_event.name());
+    if (it == evaluationContext.end()) {
+        OPENVINO_THROW("No result_event provided for OpenCL execution");
+    }
+    cl_event* ev = reinterpret_cast<cl_event*>(it->second.as<void**>());
+    *ev = ctx.lastEvent;
+}
+
+gc::gpu::OclContext MLIREvaluateGcGPU::build_ocl_context(const ov::EvaluationContext& evaluationContext) {
+    auto it = evaluationContext.find(ov::intel_gpu::ocl_queue.name());
+    if (it == evaluationContext.end()) {
+        OPENVINO_THROW("No queue provided for OpenCL execution");
+    }
+    cl_command_queue queue = reinterpret_cast<cl_command_queue>(it->second.as<void*>());
+
+    uint32_t waitListLen = 0;
+    std::vector<void*> waitList;
+
+    it = evaluationContext.find(ov::internal::mlir_meta::wait_list.name());
+    if (it != evaluationContext.end()) {
+        waitList = it->second.as<std::vector<void*>>();
+        waitListLen = waitList.size();
+    }
+
+    return gc::gpu::OclContext(module->runtime, queue, /*preserveOrder=*/waitListLen > 0,
```

Comment:
We should always set `true` here in case of out-of-order queue. If waitList is empty, we will get false here.